packages/flutter/lib/shell.dart

```diff
@@ -0,0 +1,99 @@
+    if (!servicesHandle.isValid || !exposedServicesHandle.isValid)
+      return null;
+    mojom.ServiceProviderProxy services = new mojom.ServiceProviderProxy.fromHandle(servicesHandle);
+    mojom.ServiceProviderStub exposedServices = new mojom.ServiceProviderStub.fromHandle(exposedServicesHandle);
+    return new ApplicationConnection(exposedServices, services);
+  }
+  final ApplicationConnection _embedderConnection = _initEmbedderConnection();
+
+  /// Attempts to connect to an application via the Mojo shell.
+  ApplicationConnection connectToApplication(String url) {
+    if (_shell == null)
+      return null;
+    mojom.ServiceProviderProxy services = new mojom.ServiceProviderProxy.unbound();
+    mojom.ServiceProviderStub exposedServices = new mojom.ServiceProviderStub.unbound();
+    _shell.connectToApplication(url, services, exposedServices);
+    return new ApplicationConnection(exposedServices, services);
+  }
+
+  /// Set this to intercept calls to [connectToService()] and supply an
+  /// alternative implementation of a service (for example, a mock for testing).
+  OverrideConnectToService overrideConnectToService;
+
+  /// Attempts to connect to a service implementing the interface for the given proxy.
+  /// If an application URL is specified, the service will be requested from that application.
+  /// Otherwise, it will be requested from the embedder (the Flutter engine).
+  void connectToService(String url, bindings.ProxyBase proxy) {
+    if (overrideConnectToService != null && overrideConnectToService(url, proxy))
+      return;
+    if (url == null || _shell == null) {
+      // If the application URL is null, it means the service to connect
+      // to is one provided by the embedder.
+      // If the applircation URL isn't null but there's no shell, then
+      // ask the embedder in case it provides it. (For example, if you're
+      // running on Android without the Mojo shell, then you can obtain
+      // the media service from the embedder directly, instead of having
+      // to ask the media application for it.)
+      // This makes it easier to write an application that works both
+      // with and without a Mojo environment.
+      _embedderConnection?.requestService(proxy);
+      return;
+    }
+    mojom.ServiceProviderProxy services = new mojom.ServiceProviderProxy.unbound();
+    _shell.connectToApplication(url, services, null);
+    core.MojoMessagePipe pipe = new core.MojoMessagePipe();
+    proxy.impl.bind(pipe.endpoints[0]);
+    services.ptr.connectToService(proxy.serviceName, pipe.endpoints[1]);
+    services.close();
+  }
+
+  /// Registers a service to expose to the embedder.
+  void provideService(String interfaceName, ServiceFactory factory) {
+    _embedderConnection?.provideService(interfaceName, factory);
+  }
+}
+
+/// The singleton object that manages connections with embedder-provided services.
+MojoShell get shell => MojoShell.instance;
```

Comment:
Do we want people to say shell.foo rather than MojoShell.instance.foo? (I know this existed already, just looking at this reminded me about it.)